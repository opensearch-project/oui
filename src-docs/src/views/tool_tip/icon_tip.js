/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { Fragment } from 'react';

import {
  OuiCheckbox,
  OuiFlexGroup,
  OuiFlexItem,
  OuiIconTip,
  OuiSpacer,
  OuiText,
  OuiCode,
} from '../../../../src/components';

export default () => (
  <Fragment>
    <OuiFlexGroup alignItems="center" gutterSize="s" responsive={false}>
      <OuiFlexItem grow={false}>
        <OuiCheckbox
          id="explainedCheckbox"
          label="Use source maps"
          onChange={() => {}}
        />
      </OuiFlexItem>

      <OuiFlexItem grow={false}>
        <OuiIconTip
          content="Source maps allow browser dev tools to map minified code to the original source code"
          position="right"
        />
      </OuiFlexItem>
    </OuiFlexGroup>

    <OuiSpacer />

    <OuiIconTip
      aria-label="Warning"
      size="xl"
      type="alert"
      color="warning"
      content="Proceed with caution"
    />

    <OuiSpacer />

    <OuiText>
      <p>
        Pass a position utility class via <OuiCode>iconProps</OuiCode> to shift
        for better alignment.
        <OuiIconTip
          type="iInCircle"
          color="subdued"
          content={
            <span>
              This was passed <OuiCode>.oui-alignTop</OuiCode>
            </span>
          }
          iconProps={{
            className: 'oui-alignTop',
          }}
        />
      </p>
    </OuiText>
  </Fragment>
);
