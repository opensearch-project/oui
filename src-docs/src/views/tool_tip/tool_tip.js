/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';

import {
  OuiIcon,
  OuiToolTip,
  OuiLink,
  OuiText,
  OuiFieldText,
  OuiSpacer,
  OuiButton,
} from '../../../../src/components';

export default () => (
  <div>
    <OuiText>
      <p>
        This tooltip appears on the{' '}
        <OuiToolTip position="top" content="Here is some tooltip text">
          <OuiLink href="#">top</OuiLink>
        </OuiToolTip>
      </p>

      <p>
        This tooltip appears on the{' '}
        <OuiToolTip
          position="left"
          title="Tooltip titles are optional"
          content="Here is some tooltip text. Lets add some more content to see how it wraps.">
          <OuiLink href="#">left</OuiLink>
        </OuiToolTip>{' '}
        and includes the optional title.
      </p>

      <p>
        This tooltip appears on the{' '}
        <OuiToolTip position="right" content="Here is some tooltip text">
          <OuiLink href="#">right</OuiLink>
        </OuiToolTip>
      </p>

      <p>
        This tooltip has a long delay because it might be in a repeatable
        component{' '}
        <OuiToolTip delay="long" content="Here is some tooltip text">
          <OuiLink href="#">wink</OuiLink>
        </OuiToolTip>
      </p>

      <p>
        This tooltip appears on the bottom of this icon:{' '}
        <OuiToolTip position="bottom" content="Here is some tooltip text">
          <OuiIcon tabIndex="0" type="alert" aria-label="Alert" />
        </OuiToolTip>
      </p>
    </OuiText>

    <OuiSpacer />

    <p>
      <OuiToolTip
        position="top"
        content="Here is some tooltip text"
        display="block">
        <OuiButton fullWidth>
          I am a block level tooltip, applied to a button with fullWidth
        </OuiButton>
      </OuiToolTip>
    </p>

    <OuiSpacer />

    <OuiToolTip position="right" content="Works on anything">
      <OuiFieldText
        placeholder="Hover over me"
        aria-label="ToolTip appears on hover"
      />
    </OuiToolTip>

    <OuiSpacer />

    <OuiToolTip
      position="top"
      content={
        <p>
          Works on any kind of element &mdash; buttons, inputs, you name it!
        </p>
      }>
      <OuiButton onClick={() => {}}>Hover me</OuiButton>
    </OuiToolTip>
  </div>
);
