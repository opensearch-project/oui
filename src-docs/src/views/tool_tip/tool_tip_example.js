/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';

import { renderToHtml } from '../../services';

import { GuideSectionTypes } from '../../components';

import {
  OuiCallOut,
  OuiCode,
  OuiToolTip,
  OuiIconTip,
  OuiSpacer,
  OuiTitle,
} from '../../../../src/components';
import toolTipConfig from './playground';

import ToolTip from './tool_tip';
const toolTipSource = require('!!raw-loader!./tool_tip');
const toolTipHtml = renderToHtml(ToolTip);
const tooltipSnippet = [
  `<OuiToolTip position="top" content="Tooltip text">
  <!-- An inline element to trigger the tooltip -->
</OuiToolTip>
`,
  `<OuiToolTip title="Tooltip title" content="Tooltip text">
  <!-- An inline element to trigger the tooltip -->
</OuiToolTip>
`,
  `<OuiToolTip content="A tooltip with a long delay" delay="long">
  <!-- An inline element to trigger the tooltip -->
</OuiToolTip>
`,
];

import IconTip from './icon_tip';
const infoTipSource = require('!!raw-loader!./icon_tip');
const infoTipHtml = renderToHtml(IconTip);
const infoTipSnippet = `<OuiIconTip
  content="Tooltip text for the icon"
  position="top"
  type="iInCircle"
/>
`;

export const ToolTipExample = {
  title: 'Tooltip',
  intro: (
    <OuiCallOut title="OuiToolTip only applies to inline elements">
      <p>
        OuiToolTip wraps its children in a span element, so if you pass in a
        block-level child (e.g. a div) the resulting DOM will be in violation of
        the HTML5 spec.
      </p>
    </OuiCallOut>
  ),
  sections: [
    {
      text: (
        <>
          <p>
            Wrap <strong>OuiToolTip</strong> around any item that you need a
            tooltip for. The <OuiCode>position</OuiCode> prop will take a
            suggested position, but will change it if the tooltip gets too close
            to the edge of the screen.
          </p>

          <OuiTitle size="m">
            <h2>Applying tooltips to custom components</h2>
          </OuiTitle>

          <p>
            Internally, <strong>OuiToolTip</strong> applies{' '}
            <OuiCode>onFocus</OuiCode>, <OuiCode>onBlur</OuiCode>,{' '}
            <OuiCode>onMouseOver</OuiCode>, and <OuiCode>onMouseOut</OuiCode>{' '}
            props to whatever you pass as <OuiCode>children</OuiCode>. If you
            pass in a custom component, then you&rsquo;ll need to make sure
            these props are applied to the root element rendered by your
            component. The best way to do that is to follow{' '}
            <a href="https://github.com/opensearch-project/oui/blob/main/wiki/component-design.md#pass-through-props">
              OUI&rsquo;s guidelines on pass-through props
            </a>
            .
          </p>

          <OuiCallOut
            iconType="accessibility"
            color="warning"
            title={
              <>
                Anchoring a tooltip to a non-interactive element makes it
                difficult for keyboard-only and screen reader users to read.
              </>
            }
          />

          <OuiSpacer size="l" />

          <OuiCallOut
            iconType="accessibility"
            color="warning"
            title={
              <>
                Putting anything other than plain text in a tooltip is lost on
                screen readers. Consider switching to{' '}
                <Link to="/layout/popover">
                  <strong>OuiPopover</strong>
                </Link>{' '}
                if you need more content inside a tooltip.
              </>
            }
          />
        </>
      ),
      source: [
        {
          type: GuideSectionTypes.JS,
          code: toolTipSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: toolTipHtml,
        },
      ],

      props: { OuiToolTip },
      snippet: tooltipSnippet,
      demo: <ToolTip />,
    },
    {
      title: 'IconTip',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: infoTipSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: infoTipHtml,
        },
      ],
      text: (
        <Fragment>
          <p>
            You can use <strong>OuiIconTip</strong> to explain options, other
            controls, or entire parts of the user interface. When possible,
            surface explanations inline within the UI, and only hide them behind
            a <strong>OuiIconTip</strong> as a last resort.
          </p>
          <p>
            It accepts all the same props as <strong>OuiToolTip</strong>. For
            convenience, you can also specify optional icon{' '}
            <OuiCode>size</OuiCode>, <OuiCode>type</OuiCode>, and
            <OuiCode>color</OuiCode> props.
          </p>
        </Fragment>
      ),
      props: { OuiToolTip, OuiIconTip },
      snippet: infoTipSnippet,
      demo: <IconTip />,
    },
  ],
  playground: toolTipConfig,
};
