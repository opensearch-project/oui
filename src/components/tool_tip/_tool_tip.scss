/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * 1. Shift arrow 1px more than half its size to account for border radius
 */

.ouiToolTip {
  @include ouiToolTipStyle;
  @include ouiToolTipAnimation('top');
  position: absolute;
  opacity: 0;

  // Custom sizing
  $arrowSize: $ouiSizeM;
  $arrowPlusSize: (calc($arrowSize/2) + 1px) * -1; /* 1 */
  $arrowMinusSize: (calc($arrowSize/2) - 1px) * -1; /* 1 */

  .ouiToolTip__arrow {
    content: '';
    position: absolute;
    transform-origin: center;
    border-radius: 2px;
    background-color: $ouiTooltipBackgroundColor;
    width: $arrowSize;
    height: $arrowSize;

    transform: translateY($arrowPlusSize) rotateZ(45deg); /* 1 */
  }

  // Positions the arrow and animates in from the same side.
  &.ouiToolTip--right {
    animation-name: ouiToolTipRight;

    .ouiToolTip__arrow {
      transform: translateX($arrowMinusSize) rotateZ(45deg); /* 1 */
    }
  }

  &.ouiToolTip--bottom {
    animation-name: ouiToolTipBottom;

    .ouiToolTip__arrow {
      transform: translateY($arrowMinusSize) rotateZ(45deg); /* 1 */
    }
  }

  &.ouiToolTip--left {
    animation-name: ouiToolTipLeft;

    .ouiToolTip__arrow {
      transform: translateX($arrowPlusSize) rotateZ(45deg); /* 1 */
    }
  }

  .ouiToolTip__title {
    @include ouiToolTipTitle;
  }
}

.ouiToolTipAnchor {
  display: inline-block;

  // disabled elements don't fire mouse events which means leaving a disabled element
  // wouldn't trigger the onMouseOut and hide the tooltip; disabling pointer events
  // on disabled elements means any mouse events remain handled by parent elements
  // https://jakearchibald.com/2017/events-and-disabled-form-fields/
  *[disabled] {
    pointer-events: none;
  }

  &.ouiToolTipAnchor--displayBlock {
    display: block;
  }
}

// Keyframes to animate in the tooltip.
@keyframes ouiToolTipTop {
  0% {
    opacity: 0;
    transform: translateY(-$ouiSize);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}

@keyframes ouiToolTipBottom {
  0% {
    opacity: 0;
    transform: translateY($ouiSize);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}

@keyframes ouiToolTipLeft {
  0% {
    opacity: 0;
    transform: translateX(-$ouiSize);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}

@keyframes ouiToolTipRight {
  0% {
    opacity: 0;
    transform: translateX($ouiSize);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}
