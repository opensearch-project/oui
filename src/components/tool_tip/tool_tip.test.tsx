/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { requiredProps } from '../../test';
import { OuiToolTip } from './tool_tip';

describe('OuiToolTip', () => {
  test('is rendered', () => {
    const { container } = render(
      <OuiToolTip title="title" id="id" content="content" {...requiredProps}>
        <button>Trigger</button>
      </OuiToolTip>
    );

    expect(container).toMatchSnapshot();
  });

  test('shows tooltip on focus', () => {
    jest.useFakeTimers();

    render(
      <OuiToolTip title="title" id="id" content="content" {...requiredProps}>
        <button data-test-subj="trigger">Trigger</button>
      </OuiToolTip>
    );

    // Initially tooltip should not be visible
    expect(screen.queryByText('title')).not.toBeInTheDocument();

    // Focus the trigger button
    act(() => {
      fireEvent.focus(screen.getByTestId('trigger'));
    });

    // Advance timers to trigger the tooltip display
    act(() => {
      jest.advanceTimersByTime(260); // wait for showToolTip setTimeout
    });

    // Now the tooltip should be visible
    expect(screen.getByText('title')).toBeInTheDocument();
    expect(screen.getByText('content')).toBeInTheDocument();

    // Take a snapshot of the body which should include the portal
    expect(document.body).toMatchSnapshot();

    jest.useRealTimers();
  });

  test('display prop renders block', () => {
    const { container } = render(
      <OuiToolTip
        title="title"
        id="id"
        content="content"
        {...requiredProps}
        display="block">
        <button>Trigger</button>
      </OuiToolTip>
    );

    expect(container).toMatchSnapshot();
  });
});
