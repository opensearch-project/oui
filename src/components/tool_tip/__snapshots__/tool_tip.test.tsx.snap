// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiToolTip display prop renders block 1`] = `
<span
  class="ouiToolTipAnchor ouiToolTipAnchor--displayBlock"
>
  <button>
    Trigger
  </button>
</span>
`;

exports[`OuiToolTip is rendered 1`] = `
<span
  class="ouiToolTipAnchor"
>
  <button>
    Trigger
  </button>
</span>
`;

exports[`OuiToolTip shows tooltip on focus 1`] = `
<span
  class="ouiToolTipAnchor"
>
  <button
    data-test-subj="trigger"
  >
    Trigger
  </button>
</span>
`;
